src/lib.rs

```diff
@@ -33,6 +37,51 @@ pub fn links(bytes: &[u8], links: &mut impl Extend<Cid>) -> Result<(), Error> {
     Ok(())
 }
 
+/// DAG-PB implementation of ipld-core's `Codec` trait.
+pub struct DagPbCodec;
+
+impl Codec<Ipld> for DagPbCodec {
+    const CODE: u64 = 0x70;
+    type Error = Error;
+
+    fn decode<R: BufRead>(mut reader: R) -> Result<Ipld, Self::Error> {
+        let mut bytes = Vec::new();
+        reader.read_to_end(&mut bytes)?;
+        crate::to_ipld(&bytes)
+    }
+
+    fn encode<W: Write>(mut writer: W, data: &Ipld) -> Result<(), Self::Error> {
+        let bytes = crate::from_ipld(data)?;
+        Ok(writer.write_all(&bytes)?)
+    }
+}
+
+impl Links for DagPbCodec {
+    type LinksError = Error;
+
+    fn links(data: &[u8]) -> Result<impl Iterator<Item = Cid>, Self::LinksError> {
+        let mut links = Vec::new();
+        crate::links(data, &mut links)?;
+        Ok(links.into_iter())
+    }
+}
+
+impl Codec<PbNode> for DagPbCodec {
+    const CODE: u64 = 0x70;
+    type Error = Error;
+
+    fn decode<R: BufRead>(mut reader: R) -> Result<PbNode, Self::Error> {
+        let mut bytes = Vec::new();
+        reader.read_to_end(&mut bytes)?;
+        PbNode::from_bytes(Bytes::copy_from_slice(&bytes))
+    }
+
+    fn encode<W: Write>(mut writer: W, data: &PbNode) -> Result<(), Self::Error> {
+        let bytes = data.clone().into_bytes();
```

Comment:
Can be fixed later, but the underlying `into_bytes` writes into a writer.

Also, it would be _nice_ if we could avoid cloning everything.

Reply:
Yes, it's tricky. The underlying thing is a writer, but the `into_bytes()` makes sure the link ordering is correct. As it changes the ordering of the actual object is the reason why the cloning is needed.

Yes, it could be improved, but my hope would be that no one is really using DAG-PB anyway and if, they would be open to improve it in case it's a problem.

Reply:
Honestly, I'd say:

1. Sorting links is the user's problem and the format shouldn't handle it.
2. We should actually provide multiple representations for DagPB: One that treats links as a list (letting the user sort them) and one that treats it as a map.

But let's punt that till later.

Reply:
IIRC, it's doing what the other implementations (JS and Go) are doing. We tried to make it as strict as possible to make sure the same content leads to the same hash.

Ideally we go the route of those implementations and don't even treat it as Protocol Buffers, but kind of its own custom format we parse manually.